{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%run ipy_setup.py\n",
      "%matplotlib inline\n",
      "rcParams['figure.figsize'] = (30, 10)\n",
      "rcParams['patch.edgecolor'] = 'blue'\n",
      "rcParams['figure.dpi'] = 300\n",
      "rcParams['font.size'] = 14\n",
      "rcParams['font.family'] = 'sans-serif'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "annovar_asw = pd.read_csv(\"/data/projects_gibbons/home/neffra/compaln_vcf_afrgh38_hg38_ASW_092214/asw_only.avinput.variant_function\", sep='\\t', header=None)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "annovar_aswr = pd.read_csv(\"/data/projects_gibbons/home/neffra/compaln_vcf_afrgh38_hg38_ASW_092214/asw_only_repeats.avinput.variant_function\", sep='\\t', header=None)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 46
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "annovar_com = pd.read_csv(\"/data/projects_gibbons/home/neffra/compaln_vcf_afrgh38_hg38_ASW_092214/com_only.avinput.variant_function\",sep=\"\\t\", header=None)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "for name, group in annovar_asw.groupby(annovar_asw[0]):\n",
      "    print name, len(group)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "UTR3 869\n",
        "UTR5 80\n",
        "downstream 1278\n",
        "exonic 584\n",
        "intergenic 135392\n",
        "intronic 49178\n",
        "ncRNA_UTR3 17\n",
        "ncRNA_UTR5 3\n",
        "ncRNA_exonic 494\n",
        "ncRNA_intronic 7949\n",
        "ncRNA_splicing 4\n",
        "splicing 9\n",
        "upstream 1092\n",
        "upstream;downstream 75\n"
       ]
      }
     ],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "genes = ''\n",
      "for name, group in annovar_aswr.groupby(annovar_aswr[0]):\n",
      "    print name, len(group)\n",
      "    genes = group.groupby(group[1].map(lambda x: x.split('(')[0])).size()\n",
      "    genes.sort(ascending=False)\n",
      "    print genes[0:10], '\\n'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "UTR3 369\n",
        "1\n",
        "PTCD3     6\n",
        "ZNF483    5\n",
        "CALU      5\n",
        "CD28      5\n",
        "CDH12     5\n",
        "PAX4      5\n",
        "ZNF783    4\n",
        "HES2      4\n",
        "RBM47     4\n",
        "PREPL     4\n",
        "dtype: int64 \n",
        "\n",
        "UTR5 25\n",
        "1\n",
        "GPR85       3\n",
        "CHRNA2      3\n",
        "SLC25A25    2\n",
        "PTPN7       2\n",
        "NOTCH2NL    2\n",
        "NOS1AP      2\n",
        "MYOZ3       2\n",
        "DMXL1       2\n",
        "CA14        2\n",
        "ZHX2        1\n",
        "dtype: int64 \n",
        "\n",
        "downstream 539\n",
        "1\n",
        "HLA-DQB1     6\n",
        "MUT          5\n",
        "LINC01447    5\n",
        "GAD1         5\n",
        "LINC00862    5\n",
        "FAM120B      4\n",
        "FAM231A      4\n",
        "CAV3         4\n",
        "SPINK2       4\n",
        "PANDAR       4\n",
        "dtype: int64 \n",
        "\n",
        "exonic 11\n",
        "1\n",
        "TNRC18       2\n",
        "SUPT20HL2    2\n",
        "SHROOM4      2\n",
        "IGSF3        2\n",
        "ODF1         1\n",
        "LRSAM1       1\n",
        "ADRA2B       1\n",
        "dtype: int64 \n",
        "\n",
        "intergenic 51447\n",
        "1\n",
        "NONE            1380\n",
        "TTTY23B          456\n",
        "SHOX             176\n",
        "LOC102723427     165\n",
        "CWH43            156\n",
        "STIM2            144\n",
        "CSMD1            137\n",
        "ASMT             136\n",
        "LOC100506207     135\n",
        "ELTD1            128\n",
        "dtype: int64"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " \n",
        "\n",
        "intronic 28583\n",
        "1\n",
        "CNTNAP2    172\n",
        "CSMD1      131\n",
        "PTPRD      130\n",
        "DMD        124\n",
        "DPP10       98\n",
        "LRP1B       92\n",
        "FHIT        87\n",
        "SGCZ        78\n",
        "PARK2       78\n",
        "KCNIP4      78\n",
        "dtype: int64 \n",
        "\n",
        "ncRNA_UTR3 7\n",
        "1\n",
        "RAET1E     3\n",
        "FAM167A    2\n",
        "BCL2L15    2\n",
        "dtype: int64 \n",
        "\n",
        "ncRNA_exonic 78\n",
        "1\n",
        "LOC407835       4\n",
        "LOC101928068    3\n",
        "C1orf180        3\n",
        "MEIS1-AS3       3\n",
        "LINC-PINT       2\n",
        "LOC101927619    2\n",
        "LOC101927391    2\n",
        "TRG-AS1         2\n",
        "LOC100506730    2\n",
        "LOC100287015    2\n",
        "dtype: int64 \n",
        "\n",
        "ncRNA_intronic 3683\n",
        "1\n",
        "PTCHD1-AS       48\n",
        "CASC15          47\n",
        "DIRC3           41\n",
        "MIR5096         34\n",
        "LOC441242       34\n",
        "LOC100506457    32\n",
        "LOC101927967    31\n",
        "LOC729732       27\n",
        "STEAP2-AS1      27\n",
        "MIR548F3        27\n",
        "dtype: int64 \n",
        "\n",
        "upstream 358\n",
        "1\n",
        "RARS            7\n",
        "LOC101928154    6\n",
        "CAV2            6\n",
        "G0S2            5\n",
        "HLA-DQB1        4\n",
        "LOC344887       4\n",
        "LOC100132735    4\n",
        "LINC01304       4\n",
        "MME             4\n",
        "SFT2D3          4\n",
        "dtype: int64 \n",
        "\n",
        "upstream;downstream 16\n",
        "1\n",
        "MIR4477B;MIR4477A      8\n",
        "MTFR1L;LOC646471       2\n",
        "MIR4477A;MIR4477B      2\n",
        "MIR3688-1;MIR3688-2    2\n",
        "LCN8;LCN15             2\n",
        "dtype: int64 \n",
        "\n"
       ]
      }
     ],
     "prompt_number": 47
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "genes = ''\n",
      "for name, group in annovar_asw.groupby(annovar_asw[0]):\n",
      "    print name, len(group)\n",
      "    genes = group.groupby(group[1].map(lambda x: x.split('(')[0])).size()\n",
      "    genes.sort(ascending=False)\n",
      "    print genes[0:10], '\\n'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "UTR3 869\n",
        "1\n",
        "GXYLT1      12\n",
        "HLA-DRB5    12\n",
        "PLCXD1      12\n",
        "HLA-DRB1    11\n",
        "ZNF680       9\n",
        "CISD2        8\n",
        "EIF4EBP1     6\n",
        "LCLAT1       5\n",
        "PCBP2        5\n",
        "FAM136A      5\n",
        "dtype: int64 \n",
        "\n",
        "UTR5 80\n",
        "1\n",
        "HLA-DRB1    2\n",
        "APOD        2\n",
        "ARL14       2\n",
        "FANK1       2\n",
        "SLC25A25    2\n",
        "CHRNA2      2\n",
        "TRIM15      2\n",
        "ZNF331      1\n",
        "FAM227B     1\n",
        "GABRA2      1\n",
        "dtype: int64 \n",
        "\n",
        "downstream 1278\n",
        "1\n",
        "HLA-DRB1      35\n",
        "DUX4,DUX4L    34\n",
        "HLA-DRB6      32\n",
        "DUX4L4        20\n",
        "PRSS2         17\n",
        "PRSS3         11\n",
        "PRSS1         11\n",
        "LOC554223      8\n",
        "FAM86B3P       7\n",
        "RGMA           7\n",
        "dtype: int64 \n",
        "\n",
        "exonic 584\n",
        "1\n",
        "MUC3A                                    72\n",
        "MUC6                                     33\n",
        "DUX4L4                                   32\n",
        "DUX2,DUX4,DUX4L2,DUX4L3,DUX4L5,DUX4L6    27\n",
        "MUC16                                    24\n",
        "ZNF880                                   20\n",
        "HLA-DRB5                                 13\n",
        "HLA-DRB1                                 12\n",
        "MUC4                                     11\n",
        "PRAMEF7                                  10\n",
        "dtype: int64 \n",
        "\n",
        "intergenic 135392\n",
        "1\n",
        "NONE            18638\n",
        "TTTY23B          8293\n",
        "CWH43            3612\n",
        "KCNJ18           2121\n",
        "MLLT10P1         2037\n",
        "EPHA3            1718\n",
        "ACTR3BP5         1486\n",
        "HAVCR1P1         1168\n",
        "LOC102723769      790\n",
        "DUXAP8            748\n",
        "dtype: int64"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " \n",
        "\n",
        "intronic 49178\n",
        "1\n",
        "HLA-DRB1    261\n",
        "CSMD1       211\n",
        "RBFOX1      170\n",
        "ZNF717      165\n",
        "TTC34       162\n",
        "CNTNAP2     144\n",
        "DHRSX       138\n",
        "HLA-DRB5    136\n",
        "PTPRD       127\n",
        "DMD         101\n",
        "dtype: int64"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " \n",
        "\n",
        "ncRNA_UTR3 17\n",
        "1\n",
        "FAM167A         2\n",
        "BDNF            2\n",
        "ZRANB2          1\n",
        "TTC28           1\n",
        "RAET1E          1\n",
        "NALCN           1\n",
        "MYOCD           1\n",
        "MACC1           1\n",
        "JAKMIP2         1\n",
        "GSTT2,GSTT2B    1\n",
        "dtype: int64 \n",
        "\n",
        "ncRNA_UTR5 3\n",
        "1\n",
        "TMEM220    1\n",
        "SNCA       1\n",
        "BDNF       1\n",
        "dtype: int64 \n",
        "\n",
        "ncRNA_exonic 494\n",
        "1\n",
        "LINC00202-2    36\n",
        "ROCK1P1        30\n",
        "SNORD3A        22\n",
        "DUX4L          17\n",
        "DBET           13\n",
        "FAM182A        11\n",
        "HLA-DRB6       10\n",
        "ASMTL-AS1      10\n",
        "LINC00202-1     9\n",
        "LINC00960       9\n",
        "dtype: int64 \n",
        "\n",
        "ncRNA_intronic 7949\n",
        "1\n",
        "ROCK1P1         217\n",
        "TEKT4P2         161\n",
        "HLA-DRB6        131\n",
        "PRH1-PRR4       126\n",
        "LOC441242        85\n",
        "FAM27A           82\n",
        "LOC101928804     74\n",
        "ANKRD30BL        63\n",
        "CASC15           60\n",
        "FAM230C          57\n",
        "dtype: int64 \n",
        "\n",
        "ncRNA_splicing 4\n",
        "1\n",
        "LOC283177       1\n",
        "LOC102723769    1\n",
        "FAM182A         1\n",
        "DDX12P          1\n",
        "dtype: int64 \n",
        "\n",
        "splicing 9\n",
        "1\n",
        "MROH2B     2\n",
        "PPFIA2     1\n",
        "PABPC1     1\n",
        "NUSAP1     1\n",
        "MUC16      1\n",
        "GSTT2      1\n",
        "FANCD2     1\n",
        "FAM104B    1\n",
        "dtype: int64 \n",
        "\n",
        "upstream 1092\n",
        "1\n",
        "ROCK1P1      77\n",
        "HLA-DRB5     48\n",
        "HLA-DRB1     15\n",
        "LOC388882    15\n",
        "MIR663AHG    12\n",
        "MGC39584     11\n",
        "ELP3          7\n",
        "HLA-A         7\n",
        "TLR8-AS1      7\n",
        "FAM27E2       6\n",
        "dtype: int64 \n",
        "\n",
        "upstream;downstream 75\n",
        "1\n",
        "MIR4477B;MIR4477A         33\n",
        "MIR4477A;MIR4477B         21\n",
        "MTFR1L;LOC646471           3\n",
        "SNAR-B1,SNAR-B2;SNAR-D     2\n",
        "RNU2-1;RNU2-1              2\n",
        "ZNF70;VPREB3               1\n",
        "MIR3688-1;MIR3688-2        1\n",
        "HAPLN4;TM6SF2              1\n",
        "HBBP1;BGLT3                1\n",
        "HIST3H2BB;HIST3H2A         1\n",
        "dtype: int64 \n",
        "\n"
       ]
      }
     ],
     "prompt_number": 44
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "for name, group in annovar_com.groupby(annovar_com[0]):\n",
      "    print name, len(group)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "UTR3 529\n",
        "UTR5 85\n",
        "downstream 640\n",
        "exonic 558\n",
        "intergenic 188513\n",
        "intronic 25894\n",
        "ncRNA_UTR3 21\n",
        "ncRNA_UTR5 6\n",
        "ncRNA_exonic 363\n",
        "ncRNA_intronic 4502\n",
        "ncRNA_splicing 1\n",
        "splicing 9\n",
        "upstream 752\n",
        "upstream;downstream 30\n"
       ]
      }
     ],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "genes = ''\n",
      "for name, group in annovar_com.groupby(annovar_com[0]):\n",
      "    print name, len(group)\n",
      "    genes = group.groupby(group[1].map(lambda x: x.split('(')[0])).size()\n",
      "    genes.sort(ascending=False)\n",
      "    print genes[0:10], '\\n'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "UTR3 529\n",
        "1\n",
        "HLA-DRB5    19\n",
        "PTGES3      16\n",
        "HLA-DRB1    10\n",
        "GXYLT1       7\n",
        "CTBP2        5\n",
        "HLA-B        5\n",
        "MST1L        4\n",
        "MMS22L       4\n",
        "ZNF780A      4\n",
        "GLIPR1       4\n",
        "dtype: int64 \n",
        "\n",
        "UTR5 85\n",
        "1\n",
        "HLA-DRB1    17\n",
        "GSDMC        9\n",
        "CGB2         4\n",
        "FANK1        3\n",
        "ADCY4        2\n",
        "ARL14        2\n",
        "TRIM15       2\n",
        "POU5F1       2\n",
        "CRYZ         2\n",
        "ZNF761       1\n",
        "dtype: int64 \n",
        "\n",
        "downstream 640\n",
        "1\n",
        "HLA-DRB6     26\n",
        "HLA-DRB1     23\n",
        "HLA-DQB2     12\n",
        "KRTAP9-4      9\n",
        "FAXC          9\n",
        "HLA-A         8\n",
        "HLA-DRB5      8\n",
        "PRSS1         7\n",
        "LOC554223     6\n",
        "DUX4L4        6\n",
        "dtype: int64 \n",
        "\n",
        "exonic 558\n",
        "1\n",
        "MUC3A                                    103\n",
        "MUC6                                      39\n",
        "MUC16                                     38\n",
        "AHNAK2                                    16\n",
        "DUX4L4                                    13\n",
        "HLA-DRB5                                  12\n",
        "HLA-DQB2                                  10\n",
        "HLA-DRB1                                   9\n",
        "MUC4                                       8\n",
        "DUX2,DUX4,DUX4L2,DUX4L3,DUX4L5,DUX4L6      8\n",
        "dtype: int64 \n",
        "\n",
        "intergenic 188513\n",
        "1\n",
        "NONE         93180\n",
        "EMBP1        33435\n",
        "CWH43         2397\n",
        "KCNJ18        1334\n",
        "EPHA3         1216\n",
        "ACTR3BP5      1135\n",
        "MLLT10P1      1134\n",
        "ZNF716         621\n",
        "C21orf140      570\n",
        "DUXAP8         545\n",
        "dtype: int64"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " \n",
        "\n",
        "intronic 25894\n",
        "1\n",
        "HLA-DRB5    323\n",
        "MUC16       247\n",
        "TTC34       141\n",
        "CSMD1       129\n",
        "ZNF717       94\n",
        "RBFOX1       93\n",
        "SGCZ         90\n",
        "HLA-DRB1     82\n",
        "CNTNAP2      81\n",
        "DPP6         81\n",
        "dtype: int64 \n",
        "\n",
        "ncRNA_UTR3 21\n",
        "1\n",
        "ANKEF1     4\n",
        "TTC28      2\n",
        "EYA4       2\n",
        "ZNF341     1\n",
        "ZNF217     1\n",
        "ZAK        1\n",
        "SPECC1L    1\n",
        "RRP7A      1\n",
        "NREP       1\n",
        "JAKMIP2    1\n",
        "dtype: int64 \n",
        "\n",
        "ncRNA_UTR5 6\n",
        "1\n",
        "TEKT4    5\n",
        "OBSCN    1\n",
        "dtype: int64 \n",
        "\n",
        "ncRNA_exonic 363\n",
        "1\n",
        "ROCK1P1                39\n",
        "LINC00202-2            21\n",
        "SNAR-H                  9\n",
        "DBET                    9\n",
        "MIR663A                 9\n",
        "HLA-DRB6                8\n",
        "LINC00273               7\n",
        "SNORD3B-1,SNORD3B-2     7\n",
        "LOC101928804            6\n",
        "LOC101929124            6\n",
        "dtype: int64 \n",
        "\n",
        "ncRNA_intronic 4502\n",
        "1\n",
        "ROCK1P1         177\n",
        "LOC441242        95\n",
        "HLA-DRB6         88\n",
        "FAM230C          74\n",
        "PRH1-PRR4        63\n",
        "TEKT4P2          59\n",
        "ANKRD30BL        56\n",
        "LOC100288637     44\n",
        "GSTTP2           36\n",
        "LOC101928804     35\n",
        "dtype: int64"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " \n",
        "\n",
        "ncRNA_splicing 1\n",
        "1\n",
        "ZNF300P1    1\n",
        "dtype: int64 \n",
        "\n",
        "splicing 9\n",
        "1\n",
        "DUX4        2\n",
        "POU2F3      1\n",
        "MYH11       1\n",
        "MUC16       1\n",
        "FCAMR       1\n",
        "EPB41L4A    1\n",
        "CHRNA3      1\n",
        "ADAM29      1\n",
        "dtype: int64 \n",
        "\n",
        "upstream 752\n",
        "1\n",
        "ROCK1P1     85\n",
        "GSDMC       28\n",
        "LILRA1      27\n",
        "FAM230B     19\n",
        "HLA-C       14\n",
        "HLA-DRB1    14\n",
        "MIR1299     14\n",
        "MGC39584    14\n",
        "HLA-H       13\n",
        "HLA-DQB2    12\n",
        "dtype: int64 \n",
        "\n",
        "upstream;downstream 30\n",
        "1\n",
        "MIR4477B;MIR4477A                14\n",
        "MIR4477A;MIR4477B                 4\n",
        "RNU2-1;RNU2-1                     2\n",
        "TRIM54;DNAJC5G                    1\n",
        "SLC25A34;PLEKHM2                  1\n",
        "SCIMP;LOC100130950                1\n",
        "RRM1;STIM1                        1\n",
        "MIR5692C2;MIR5692A1,MIR5692A2     1\n",
        "MIR517B;MIR520D                   1\n",
        "MIR4520-1;MIR4520-2               1\n",
        "dtype: int64 \n",
        "\n"
       ]
      }
     ],
     "prompt_number": 45
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "'''\n",
      "Value\tDefault precedence\tExplanation\n",
      "exonic\t1\tvariant overlaps a coding exon\n",
      "splicing\t1\tvariant is within 2-bp of a splicing junction (use -splicing_threshold to change this)\n",
      "ncRNA\t2\tvariant overlaps a transcript without coding annotation in the gene definition (see Notes below for more explanation)\n",
      "UTR5\t3\tvariant overlaps a 5' untranslated region\n",
      "UTR3\t3\tvariant overlaps a 3' untranslated region\n",
      "intronic\t4\tvariant overlaps an intron\n",
      "upstream\t5\tvariant overlaps 1-kb region upstream of transcription start site\n",
      "downstream\t5\tvariant overlaps 1-kb region downtream of transcription end site (use -neargene to change this)\n",
      "intergenic\t6\tvariant is in intergenic region\n",
      "'''\n",
      "pass"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print len(annovar_exonic)\n",
      "for name, group in annovar_exonic.groupby(annovar_exonic[1]):\n",
      "    print name, len(group)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "8586\n",
        "frameshift deletion 6\n",
        "frameshift insertion 6\n",
        "nonframeshift deletion 13\n",
        "nonframeshift insertion 15\n",
        "nonsynonymous SNV 3896\n",
        "stopgain 18\n",
        "stoploss 5\n",
        "synonymous SNV 4571\n",
        "unknown 56\n"
       ]
      }
     ],
     "prompt_number": 15
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "'''\n",
      "Annotation\tPrecedence\tExplanation\n",
      "frameshift insertion\t1\tan insertion of one or more nucleotides that cause frameshift changes in protein coding sequence\n",
      "frameshift deletion\t2\ta deletion of one or more nucleotides that cause frameshift changes in protein coding sequence\n",
      "frameshift block substitution\t3\ta block substitution of one or more nucleotides that cause frameshift changes in protein coding sequence\n",
      "stopgain\t4\ta nonsynonymous SNV, frameshift insertion/deletion, nonframeshift insertion/deletion or block substitution that lead to the immediate creation of stop codon at the variant site. For frameshift mutations, the creation of stop codon downstream of the variant will not be counted as \"stopgain\"!\n",
      "stoploss\t5\ta nonsynonymous SNV, frameshift insertion/deletion, nonframeshift insertion/deletion or block substitution that lead to the immediate elimination of stop codon at the variant site\n",
      "nonframeshift insertion\t6\tan insertion of 3 or multiples of 3 nucleotides that do not cause frameshift changes in protein coding sequence\n",
      "nonframeshift deletion\t7\ta deletion of 3 or mutliples of 3 nucleotides that do not cause frameshift changes in protein coding sequence\n",
      "nonframeshift block substitution\t8\ta block substitution of one or more nucleotides that do not cause frameshift changes in protein coding sequence\n",
      "nonsynonymous SNV\t9\ta single nucleotide change that cause an amino acid change\n",
      "synonymous SNV\t10\ta single nucleotide change that does not cause an amino acid change\n",
      "unknown\t11\tunknown function (due to various errors in the gene structure definition in the database file)\n",
      "'''\n",
      "pass"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    }
   ],
   "metadata": {}
  }
 ]
}