{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%load /data/projects_gibbons/home/neffra/scripts/python/compare_aln_vcf.py"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!/data/projects_gibbons/home/neffra/anaconda2/bin/python\n",
      "\n",
      "import sys, getopt\n",
      "import time\n",
      "import numpy as np\n",
      "import scipy as sp\n",
      "import pandas as pd\n",
      "import pysam\n",
      "import subprocess\n",
      "import os\n",
      "\n",
      "def main(argv):\n",
      "\torig_bam = ''\n",
      "   \tfinal_bam = ''\n",
      "\tliftover = ''\n",
      "\tsites_file = ''\n",
      "\tidf = ''\n",
      "   \ttry:\n",
      "\t\topts, args = getopt.getopt(argv,\"c:a:l:s:i:g\",[\"common=\",\"alt=\", \"liftover=\", \"sites=\", \"idf=\",\"chrom=\"])\n",
      "   \texcept getopt.GetoptError:\n",
      "      \t\tprint 'compare_aln_vcf.py -c <common.bam> -a <alt.bam> -l <liftover.chain> -s <sites.tsv> -i <id> -g <chrom>'\n",
      "      \t\tsys.exit(2)\n",
      "   \tfor opt, arg in opts:\n",
      "      \t\tif opt == '-h':\n",
      "         \t\tprint 'compare_aln_vcf.py -c <common.bam> -a <alt.bam> -l <liftover.chain> -s <sites.tsv> -i <id> -g <chrom>'\n",
      "         \t\tsys.exit()\n",
      "      \t\telif opt in (\"-c\", \"--common\"):\n",
      "         \t\torig_bam = arg\n",
      "      \t\telif opt in (\"-a\", \"--alt\"):\n",
      "         \t\tfinal_bam = arg\n",
      "\t\telif opt in (\"-l\", \"--liftover\"):\n",
      "\t\t\tliftover = arg\n",
      "\t\telif opt in (\"-s\", \"--sites\"):\n",
      "\t\t\tsites_file = arg\n",
      "\t\telif opt in (\"-i\", \"--idf\"):\n",
      "\t\t\tidf = arg\n",
      "                elif opt in (\"-g\", \"--chrom\"):\n",
      "                        chrom = arg\n",
      "\t\telse:\n",
      "\t\t\tassert False, \"unhandled option\"\n",
      "\tif idf == '':\n",
      "\t\tidf = orig_bam.split(\"/\")[-1].split(\"_\")[0]\n",
      "\tcompare_alignments_vcf(orig_bam, final_bam, idf, liftover, sites_file)\n",
      "\n",
      "def compare_alignments_vcf(orig_bam, final_bam, idf, liftover, sites_file, chrom=None):\n",
      "    \n",
      "    #index original BAM file\n",
      "    #if os.path.isfile('.'.join(orig_bam.split('.')[0:-1]) + '.bai') == False:\n",
      "    subprocess.check_output('samtools index ' + orig_bam, shell=True)\n",
      "    \n",
      "    #index final BAM file\n",
      "    #if os.path.isfile('.'.join(final_bam.split('.')[0:-1]) + '.bai') == False:\n",
      "    subprocess.check_output('samtools index ' + final_bam, shell=True)\n",
      "    \n",
      "\n",
      "    '''\n",
      "    Usage: Compares detailed alignment statistics between two files. Should be\n",
      "    a comparison of alignments between the same individual.                          \n",
      "    \n",
      "    Inputs:\n",
      "        orig_bam (string)\n",
      "            The filename of a bam file aligned to the original \n",
      "            (e.g. common) reference genome. \n",
      "        final_bam (string)\n",
      "            The filename of a bam file aligned to the final \n",
      "            (e.g. pop-specific) reference.\n",
      "            genome. \n",
      "        liftover (string)\n",
      "            The filename of a liftover chain file generated during the \n",
      "            orig->final reference creation process.\n",
      "        sites_file (string)\n",
      "            The filename of the sites changed between references for \n",
      "            statistics collection. This will be formatted identical\n",
      "\t    to a VCF file.\n",
      "        idf (string)\n",
      "            This id is appended at the beginning of the outfile name. The \n",
      "            orig_bam and final_bam should have the same id. This also \n",
      "            needs to be identical to the sample name in the sites file.\n",
      "    \n",
      "    Outputs:\n",
      "        {id}-compare_d.arr\n",
      "            A listing of the collected metrics by change in each BAM file.\n",
      "    '''\n",
      "    if chrom != None:\n",
      "        out_file = final_bam + \"_\" + chrom + \"_compare_d.arr\"\n",
      "    else:\n",
      "        out_file = final_bam + \"_compare_d.arr\"\n",
      "    \n",
      "    # here for legacy\n",
      "    metrics = [{'stepper':'all','mask':2}]        # this filters for secondary/non-primary alignments\n",
      "    metrics.append({'stepper':'all','mask':3340}) # this filters for properly mapped reads\n",
      "        \n",
      "    # open BAM files given in input parameters\n",
      "    assert pysam.Samfile(orig_bam, 'rb'), 'ERROR: Cannot open orig_bam file.'\n",
      "    assert pysam.Samfile(final_bam, 'rb'), 'ERROR: Cannot open final_bam file.'\n",
      "    orig_fp = pysam.Samfile(orig_bam, 'rb')\n",
      "    final_fp = pysam.Samfile(final_bam, 'rb')\n",
      "    \n",
      "    # open sites and liftover files\n",
      "    headnum = int(subprocess.check_output('head -n 500 ' + sites_file + ' | grep \"#\" | wc -l', shell=True))# this gives us the number of lines in the header, minus the col titles\n",
      "    sites = pd.read_csv(sites_file, sep='\\t', skiprows=headnum-1)\n",
      "    lifttable = pd.read_csv(liftover, header=None, sep='\\t')\n",
      "    \n",
      "    #open output file\n",
      "    out_fp = open(out_file, \"wb\")\n",
      "    \n",
      "    # for fast retrieval - assume the input is already sorted which it has to be\n",
      "    lifttable = lifttable.set_index(lifttable[2])\n",
      "    sites = sites.set_index([sites['#CHROM'], sites['POS']])[['#CHROM','POS','REF','ALT','INFO','FORMAT',idf]]\n",
      "    \n",
      "    # some more things to prep for traversal - get list of chromosomes\n",
      "    if chrom != None:\n",
      "        chrom = [chrom]\n",
      "    else:\n",
      "        chrom = sorted(set(lifttable.index))\n",
      "    \n",
      "    # okay, we're ready to do the comparison at the sites listed - progressbar shouldn't interfere\n",
      "    \n",
      "    for cur_chrom in chrom:\n",
      "        out_arr = [] # this will hold the values until we're ready to print\n",
      "        end_length = lifttable.loc[cur_chrom].iloc[0][3] # this is the length of the contig in common ref\n",
      "        sl = list(sites.loc[cur_chrom]['POS'])\n",
      "        lift_til = 0\n",
      "        delta = 0\n",
      "        lt2 = lifttable.loc[cur_chrom] # help make things faster\n",
      "        for pos in sl:\n",
      "            orig_result = get_pileup_statistics(orig_fp, cur_chrom, pos-1) # get the result from the original BAM\n",
      "            if pos > lift_til:\n",
      "                try:\n",
      "                    delta, lift_til = liftdelta(pos-1, lt2, cur_chrom) # converts position into alt ref pos\n",
      "                except:\n",
      "\t\t    sys.stderr.write(\"WARN: couldn't lift position at \" + str(cur_chrom) + \":\" + str(pos-1) + \"\\n\")\n",
      "\t\t    continue\n",
      "            newpos = (pos-1)+delta\n",
      "            alt_result = get_pileup_statistics(final_fp, cur_chrom, newpos)\n",
      "            final_result = []\n",
      "            final_result.append([cur_chrom, pos-1])\n",
      "            final_result.append(list(sites.loc[cur_chrom].loc[pos][['REF','ALT','INFO','FORMAT',idf]]))\n",
      "            final_result.append([i for i in orig_result])\n",
      "            final_result.append([i for i in alt_result])\n",
      "            final_result = [item for sublist in final_result for item in sublist]\n",
      "            # once we're done with each chromosome, we can flush to disk (not the only solution...)\n",
      "            out_fp.write(custom_write(final_result))\n",
      "            \n",
      "    # return if everything is happy\n",
      "    sys.stderr.write(\"Completed traversal for \" + idf + \".\\n\")\n",
      "    return 0\n",
      "\n",
      "# used by compare_alignments - subroutine\n",
      "# outputs - number of aligned reads, mean mapping quality, mean insert size, mean aligned length of reads\n",
      "def get_pileup_statistics(samfile, chrom, start):\n",
      "    import numpy as np\n",
      "    result = [0, 0, 0, 0, 0] #important!\n",
      "    reached = False\n",
      "    for i in samfile.pileup(chrom, start, start+1):\n",
      "        if i.pos == start:\n",
      "            reached=True\n",
      "            mapq = []\n",
      "            insert = []\n",
      "            edit_dist = []\n",
      "            aligned_score = []\n",
      "            for p in i.pileups:\n",
      "                mapq.append(p.alignment.mapq)\n",
      "                insert.append(abs(p.alignment.tlen))\n",
      "                # average edit distance\n",
      "                tags = dict()\n",
      "                for b in p.alignment.tags:\n",
      "                    tags[b[0]] = b[1]\n",
      "                edit_dist.append(tags['NM'])\n",
      "                aligned_score.append(tags['AS'])\n",
      "            break\n",
      "    if reached==True:\n",
      "        result = (i.n, np.mean(mapq), np.mean(insert), np.mean(edit_dist), np.mean(aligned_score))\n",
      "    return result\n",
      "\n",
      "def liftdelta(pos, liftover, chrom): # going in the hg19 --> AFRG1 direction\n",
      "    match = liftover[(liftover[5] <= pos) & (liftover[6] > pos)].irow(0)\n",
      "    return match[10] - match[5], match[6]\n",
      "\n",
      "def custom_write(i):\n",
      "    result = ''\n",
      "    for s in i:\n",
      "        result += str(s) + \"\\t\"\n",
      "    result += \"\\n\"\n",
      "    return result\n",
      "\n",
      "# run the program if called from the command line\n",
      "if __name__ == \"__main__\":\n",
      "   main(sys.argv[1:])\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%run ipy_setup.py"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%run ipy_parallel.py"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "importing defaultdict from collections on engine(s)\n",
        "importing json on engine(s)\n",
        "importing sys on engine(s)\n",
        "importing numpy on engine(s)\n",
        "importing scipy on engine(s)"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "importing matplotlib.pyplot on engine(s)\n",
        "importing pandas on engine(s)\n",
        "importing rcParams from matplotlib on engine(s)\n",
        "importing matplotlib.cm on engine(s)"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "importing matplotlib on engine(s)\n",
        "importing Bio on engine(s)\n",
        "importing HTSeq on engine(s)\n"
       ]
      }
     ],
     "prompt_number": 88
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tup = (3,4,5)\n",
      "a,b,c = tup\n",
      "print a\n",
      "print b\n",
      "print c"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "3\n",
        "4\n",
        "5\n"
       ]
      }
     ],
     "prompt_number": 83
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!/data/projects_gibbons/home/neffra/anaconda2/bin/python\n",
      "\n",
      "def compare_alignments_vcf(orig_bam, final_bam, idf, liftover, sites_file, chrom=None):\n",
      "    import sys, getopt\n",
      "    import time\n",
      "    import numpy as np\n",
      "    import scipy as sp\n",
      "    import pandas as pd\n",
      "    import pysam\n",
      "    import subprocess\n",
      "    import os\n",
      "    \n",
      "    #index original BAM file\n",
      "    if os.path.isfile('.'.join(orig_bam.split('.')[0:-1]) + '.bai') == False:\n",
      "        subprocess.check_output('samtools index ' + orig_bam, shell=True)\n",
      "    \n",
      "    #index final BAM file\n",
      "    #if os.path.isfile('.'.join(final_bam.split('.')[0:-1]) + '.bai') == False:\n",
      "    #subprocess.check_output('samtools index ' + final_bam, shell=True)\n",
      "    \n",
      "\n",
      "    '''\n",
      "    Usage: Compares detailed alignment statistics between two files. Should be\n",
      "    a comparison of alignments between the same individual.                          \n",
      "    \n",
      "    Inputs:\n",
      "        orig_bam (string)\n",
      "            The filename of a bam file aligned to the original \n",
      "            (e.g. common) reference genome. \n",
      "        final_bam (string)\n",
      "            The filename of a bam file aligned to the final \n",
      "            (e.g. pop-specific) reference.\n",
      "            genome. \n",
      "        liftover (string)\n",
      "            The filename of a liftover chain file generated during the \n",
      "            orig->final reference creation process.\n",
      "        sites_file (string)\n",
      "            The filename of the sites changed between references for \n",
      "            statistics collection. This will be formatted identical\n",
      "\t    to a VCF file.\n",
      "        idf (string)\n",
      "            This id is appended at the beginning of the outfile name. The \n",
      "            orig_bam and final_bam should have the same id. This also \n",
      "            needs to be identical to the sample name in the sites file.\n",
      "    \n",
      "    Outputs:\n",
      "        {id}-compare_d.arr\n",
      "            A listing of the collected metrics by change in each BAM file.\n",
      "    '''\n",
      "    if chrom != None:\n",
      "        out_file = '/'.join(final_bam.split('/')[0:-1]) + '/' + idf + \"_\" + chrom + \"_compare_d.arr\"\n",
      "    else:\n",
      "        out_file = '/'.join(final_bam.split('/')[0:-1]) + '/' + idf + \"_compare_d.arr\"\n",
      "    \n",
      "    # here for legacy\n",
      "    metrics = [{'stepper':'all','mask':2}]        # this filters for secondary/non-primary alignments\n",
      "    metrics.append({'stepper':'all','mask':3340}) # this filters for properly mapped reads\n",
      "        \n",
      "    # open BAM files given in input parameters\n",
      "    assert pysam.Samfile(orig_bam, 'rb'), 'ERROR: Cannot open orig_bam file.'\n",
      "    assert pysam.Samfile(final_bam, 'rb'), 'ERROR: Cannot open final_bam file.'\n",
      "    orig_fp = pysam.Samfile(orig_bam, 'rb')\n",
      "    final_fp = pysam.Samfile(final_bam, 'rb')\n",
      "    \n",
      "    # open sites and liftover files\n",
      "    headnum = int(subprocess.check_output('head -n 500 ' + sites_file + ' | grep \"#\" | wc -l', shell=True))# this gives us the number of lines in the header, minus the col titles\n",
      "    sites = pd.read_csv(sites_file, sep='\\t', skiprows=headnum-1)\n",
      "    lifttable = pd.read_csv(liftover, header=None, sep='\\t')\n",
      "    \n",
      "    #open output file\n",
      "    out_fp = open(out_file, \"wb\")\n",
      "    \n",
      "    # for fast retrieval - assume the input is already sorted which it has to be\n",
      "    lifttable = lifttable.set_index(lifttable[2])\n",
      "    sites = sites.set_index([sites['#CHROM'], sites['POS']])[['#CHROM','POS','REF','ALT','INFO','FORMAT',idf]]\n",
      "    \n",
      "    # some more things to prep for traversal - get list of chromosomes\n",
      "    if chrom != None:\n",
      "        chrom = [chrom]\n",
      "    else:\n",
      "        chrom = sorted(set(lifttable.index))\n",
      "    \n",
      "    # okay, we're ready to do the comparison at the sites listed - progressbar shouldn't interfere\n",
      "    \n",
      "    for cur_chrom in chrom:\n",
      "        out_arr = [] # this will hold the values until we're ready to print\n",
      "        end_length = lifttable.loc[cur_chrom].iloc[0][3] # this is the length of the contig in common ref\n",
      "        sl = list(sites.loc[cur_chrom]['POS'])\n",
      "        lift_til = 0\n",
      "        delta = 0\n",
      "        lt2 = lifttable.loc[cur_chrom] # help make things faster\n",
      "        for pos in sl:\n",
      "            orig_result = get_pileup_statistics(orig_fp, cur_chrom, pos-1) # get the result from the original BAM\n",
      "            if pos > lift_til:\n",
      "                try:\n",
      "                    delta, lift_til = liftdelta(pos-1, lt2, cur_chrom) # converts position into alt ref pos\n",
      "                except:\n",
      "\t\t    sys.stderr.write(\"WARN: couldn't lift position at \" + str(cur_chrom) + \":\" + str(pos-1) + \"\\n\")\n",
      "\t\t    continue\n",
      "            newpos = (pos-1)+delta\n",
      "            alt_result = get_pileup_statistics(final_fp, cur_chrom, newpos)\n",
      "            final_result = []\n",
      "            final_result.append([cur_chrom, pos-1])\n",
      "            final_result.append(list(sites.loc[cur_chrom].loc[pos][['REF','ALT','INFO','FORMAT',idf]]))\n",
      "            final_result.append([i for i in orig_result])\n",
      "            final_result.append([i for i in alt_result])\n",
      "            final_result = [item for sublist in final_result for item in sublist]\n",
      "            # once we're done with each chromosome, we can flush to disk (not the only solution...)\n",
      "            out_fp.write(custom_write(final_result))\n",
      "            \n",
      "    # return if everything is happy\n",
      "    sys.stderr.write(\"Completed traversal for \" + idf + \".\\n\")\n",
      "    return 0\n",
      "\n",
      "# used by compare_alignments - subroutine\n",
      "# outputs - number of aligned reads, mean mapping quality, mean insert size, mean aligned length of reads\n",
      "def get_pileup_statistics(samfile, chrom, start):\n",
      "    import numpy as np\n",
      "    result = (0, 0, 0, 0)\n",
      "    reached = False\n",
      "    for i in samfile.pileup(chrom, start, start+1):\n",
      "        if i.pos == start:\n",
      "            reached=True\n",
      "            mapq = []\n",
      "            insert = []\n",
      "            aligned_score = []\n",
      "            for p in i.pileups:\n",
      "                mapq.append(p.alignment.mapq)\n",
      "                insert.append(abs(p.alignment.tlen))\n",
      "                # average edit distance\n",
      "                tags = dict()\n",
      "                for b in p.alignment.tags:\n",
      "                    tags[b[0]] = b[1]\n",
      "                aligned_score.append(tags['NM'])\n",
      "                aligned_score.append(tags['AS'])\n",
      "            break\n",
      "    if reached==True:\n",
      "        result = (i.n, np.mean(mapq), np.mean(insert), np.mean(aligned_score))\n",
      "    return result\n",
      "\n",
      "def liftdelta(pos, liftover, chrom): # going in the hg19 --> AFRG1 direction\n",
      "    match = liftover[(liftover[5] <= pos) & (liftover[6] > pos)].irow(0)\n",
      "    return match[10] - match[5], match[6]\n",
      "\n",
      "def custom_write(i):\n",
      "    result = ''\n",
      "    for s in i:\n",
      "        result += str(s) + \"\\t\"\n",
      "    result += \"\\n\"\n",
      "    return result"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 138
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "checkfiles = pd.read_csv('check_files_list', sep='\\t', header=None)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 79
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "checkfiles"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>2</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0 </th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19625_1.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031785</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1 </th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19625_2.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031806</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2 </th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19625_3.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031807</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3 </th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19625_4.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031808</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4 </th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19625_5.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031809</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5 </th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19625_6.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR034421</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>6 </th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19700_1.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032234</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>7 </th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19700_2.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032235</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>8 </th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19700_3.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032230</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>9 </th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19700_4.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032231</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19700_5.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032232</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>11</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19700_6.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032233</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>12</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19701_1.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031613</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>13</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19701_2.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031612</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>14</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19701_3.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031617</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>15</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19701_4.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031616</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>16</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19701_5.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031615</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>17</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19701_6.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031614</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>18</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19703_1.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032324</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>19</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19703_2.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032325</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>20</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19703_3.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR030621</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>21</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19703_4.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR030620</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>22</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19703_5.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031611</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>23</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19703_6.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031610</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>24</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19704_1.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR030625</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>25</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19704_2.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR030624</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>26</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19704_3.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR030627</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>27</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19704_4.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR030626</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>28</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19704_5.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR030629</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>29</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19704_6.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR030628</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>30</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19707_1.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032322</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>31</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19707_2.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032323</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19707_3.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032320</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>33</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19707_4.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032321</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>34</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19707_5.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032319</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>35</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19707_6.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032318</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>36</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19711_1.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032223</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>37</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19711_2.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032222</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>38</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19711_3.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032227</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>39</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19711_4.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032226</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>40</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19711_5.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032225</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>41</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19711_6.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032224</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>42</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19712_2.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031786</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>43</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19712_3.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR031810</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>44</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19712_4.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032229</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>45</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19712_5.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032228</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>46</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19712_8.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR032421</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>47</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19818_1.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR038698</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>48</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19818_2.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR038699</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>49</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19818_3.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR038700</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>50</th>\n",
        "      <td> /data/test/home/neffra/check_files/NA19818_4.r...</td>\n",
        "      <td> /data/projects_gibbons/home/neffra/1000g/SRR03...</td>\n",
        "      <td> SRR038701</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>51 rows \u00d7 3 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 38,
       "text": [
        "                                                    0                                                  1          2\n",
        "0   /data/test/home/neffra/check_files/NA19625_1.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031785\n",
        "1   /data/test/home/neffra/check_files/NA19625_2.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031806\n",
        "2   /data/test/home/neffra/check_files/NA19625_3.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031807\n",
        "3   /data/test/home/neffra/check_files/NA19625_4.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031808\n",
        "4   /data/test/home/neffra/check_files/NA19625_5.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031809\n",
        "5   /data/test/home/neffra/check_files/NA19625_6.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR034421\n",
        "6   /data/test/home/neffra/check_files/NA19700_1.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032234\n",
        "7   /data/test/home/neffra/check_files/NA19700_2.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032235\n",
        "8   /data/test/home/neffra/check_files/NA19700_3.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032230\n",
        "9   /data/test/home/neffra/check_files/NA19700_4.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032231\n",
        "10  /data/test/home/neffra/check_files/NA19700_5.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032232\n",
        "11  /data/test/home/neffra/check_files/NA19700_6.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032233\n",
        "12  /data/test/home/neffra/check_files/NA19701_1.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031613\n",
        "13  /data/test/home/neffra/check_files/NA19701_2.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031612\n",
        "14  /data/test/home/neffra/check_files/NA19701_3.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031617\n",
        "15  /data/test/home/neffra/check_files/NA19701_4.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031616\n",
        "16  /data/test/home/neffra/check_files/NA19701_5.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031615\n",
        "17  /data/test/home/neffra/check_files/NA19701_6.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031614\n",
        "18  /data/test/home/neffra/check_files/NA19703_1.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032324\n",
        "19  /data/test/home/neffra/check_files/NA19703_2.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032325\n",
        "20  /data/test/home/neffra/check_files/NA19703_3.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR030621\n",
        "21  /data/test/home/neffra/check_files/NA19703_4.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR030620\n",
        "22  /data/test/home/neffra/check_files/NA19703_5.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031611\n",
        "23  /data/test/home/neffra/check_files/NA19703_6.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031610\n",
        "24  /data/test/home/neffra/check_files/NA19704_1.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR030625\n",
        "25  /data/test/home/neffra/check_files/NA19704_2.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR030624\n",
        "26  /data/test/home/neffra/check_files/NA19704_3.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR030627\n",
        "27  /data/test/home/neffra/check_files/NA19704_4.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR030626\n",
        "28  /data/test/home/neffra/check_files/NA19704_5.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR030629\n",
        "29  /data/test/home/neffra/check_files/NA19704_6.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR030628\n",
        "30  /data/test/home/neffra/check_files/NA19707_1.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032322\n",
        "31  /data/test/home/neffra/check_files/NA19707_2.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032323\n",
        "32  /data/test/home/neffra/check_files/NA19707_3.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032320\n",
        "33  /data/test/home/neffra/check_files/NA19707_4.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032321\n",
        "34  /data/test/home/neffra/check_files/NA19707_5.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032319\n",
        "35  /data/test/home/neffra/check_files/NA19707_6.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032318\n",
        "36  /data/test/home/neffra/check_files/NA19711_1.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032223\n",
        "37  /data/test/home/neffra/check_files/NA19711_2.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032222\n",
        "38  /data/test/home/neffra/check_files/NA19711_3.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032227\n",
        "39  /data/test/home/neffra/check_files/NA19711_4.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032226\n",
        "40  /data/test/home/neffra/check_files/NA19711_5.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032225\n",
        "41  /data/test/home/neffra/check_files/NA19711_6.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032224\n",
        "42  /data/test/home/neffra/check_files/NA19712_2.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031786\n",
        "43  /data/test/home/neffra/check_files/NA19712_3.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR031810\n",
        "44  /data/test/home/neffra/check_files/NA19712_4.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032229\n",
        "45  /data/test/home/neffra/check_files/NA19712_5.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032228\n",
        "46  /data/test/home/neffra/check_files/NA19712_8.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR032421\n",
        "47  /data/test/home/neffra/check_files/NA19818_1.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR038698\n",
        "48  /data/test/home/neffra/check_files/NA19818_2.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR038699\n",
        "49  /data/test/home/neffra/check_files/NA19818_3.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR038700\n",
        "50  /data/test/home/neffra/check_files/NA19818_4.r...  /data/projects_gibbons/home/neffra/1000g/SRR03...  SRR038701\n",
        "\n",
        "[51 rows x 3 columns]"
       ]
      }
     ],
     "prompt_number": 38
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "global liftover, sites_file, chrom\n",
      "liftover = '/data/projects_gibbons/home/neffra/AFRGh38_variants/liftover_AFRGh38_090514.chain'\n",
      "orig_bam = '/data/projects_gibbons/home/neffra/NA19625_1/SRR031785_bwamem_hg38.realigned.bam'\n",
      "final_bam ='/data/projects_gibbons/home/neffra/NA19625_1/NA19625_1.realigned.bam'\n",
      "sites_file='/data/projects_gibbons/home/neffra/AFRGh38_variants/variants_set_gs_1kg_hap_hg38.vcf'\n",
      "idf = 'NA19625'\n",
      "chrom=None\n",
      "#compare_alignments_vcf(orig_bam, final_bam, liftover, sites_file, idf, chrom)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 127
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "proj_array = []\n",
      "for a,i in checkfiles.iterrows():\n",
      "    proj_array.append((i[1], i[0], i[0].split(\"/\")[-1].split(\"_\")[0], liftover, sites_file, chrom))\n",
      "compare_alignments_vcf.map(proj_array)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "CompositeError",
       "evalue": "one or more exceptions from call to method: compare_alignments_vcf\n[28:apply]: NameError: global name 'get_pileup_statistics' is not defined\n.... 50 more exceptions ...",
       "output_type": "pyerr",
       "traceback": [
        "[28:apply]: ",
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
        "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)\u001b[1;32m<string>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m",
        "\u001b[1;32m<ipython-input-135-644d1fd07c4a>\u001b[0m in \u001b[0;36mcompare_alignments_vcf\u001b[1;34m(input_comp)\u001b[0m",
        "\u001b[1;31mNameError\u001b[0m: global name 'get_pileup_statistics' is not defined",
        "",
        "... 50 more exceptions ..."
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "in sync results <function __call__ at 0x5225cf8>\n"
       ]
      }
     ],
     "prompt_number": 136
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "@direct.parallel(block=True)\n",
      "def echo(x):\n",
      "     return str(x)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 105
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "echo.map(range(5))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "in sync results <function __call__ at 0x5225cf8>\n"
       ]
      },
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 106,
       "text": [
        "['0', '1', '2', '3', '4']"
       ]
      }
     ],
     "prompt_number": 106
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "for i in pid_array:\n",
      "    print i.ready()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n",
        "False\n"
       ]
      }
     ],
     "prompt_number": 75
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# open sites and liftover files\n",
      "headnum = int(subprocess.check_output('head -n 500 ' + sites_file + ' | grep \"#\" | wc -l', shell=True))# this gives us the number of lines in the header, minus the col titles\n",
      "sites = pd.read_csv(sites_file, sep='\\t', skiprows=headnum-1)\n",
      "sites = sites.set_index(sites['#CHROM'])[['#CHROM','POS','REF','ALT','INFO','FORMAT',idf]]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "NameError",
       "evalue": "name 'cur_chrom' is not defined",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[1;32m<ipython-input-24-2ef27e783c90>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[0msites\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msites_file\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0msep\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'\\t'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mskiprows\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mheadnum\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[0msites\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0msites\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mset_index\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msites\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'#CHROM'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'#CHROM'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'POS'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'REF'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'ALT'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'INFO'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'FORMAT'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0midf\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m \u001b[0msites\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mcur_chrom\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'REF'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'ALT'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'FORMAT'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0midf\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[1;31mNameError\u001b[0m: name 'cur_chrom' is not defined"
       ]
      }
     ],
     "prompt_number": 24
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sites = sites.set_index([sites['#CHROM'], sites['POS']])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 41
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "list(sites.loc['chr20'].loc[83158][['#CHROM','POS','REF','ALT','INFO','FORMAT',idf]])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 46,
       "text": [
        "['chr20',\n",
        " 83158,\n",
        " 'C',\n",
        " 'T',\n",
        " 'AC=0;AF=0.00;AN=0;set=hapmap;set2=1kg_hapmap',\n",
        " 'GT',\n",
        " './.']"
       ]
      }
     ],
     "prompt_number": 46
    }
   ],
   "metadata": {}
  }
 ]
}