{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Python-scarecrow program for alternate sequence alignment"
     ]
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "**Pseudocode of how to set this up in Python\n",
      "\n",
      "From the Perl program description: \n",
      "\n",
      "# The program runs in two stages. First, the sub-population consensus\n",
      "# sequence for the individual is created from 1000 Genomes' data, \n",
      "# accounting for SNPs, indels, and large structural variation using \n",
      "# available genotype data. Unmapped contigs from the 1000 Genomes project\n",
      "# are assembled and by de novo sequencing and mapped to the human genome\n",
      "# by admixture mapping of linkage-associated markers present in the \n",
      "# de novo contigs and available genotype data. Next, the sub-population\n",
      "# consensus sequence is used to align the reads from the individual. \n",
      "\n",
      "REQUIRE:\n",
      "    GATK installed, with support for FastaAlternateSequence *\n",
      "    Java *\n",
      "    bwa / bowtie\n",
      "    Biopython or HTSeq library *\n",
      "    \n",
      "\n",
      "0. Inputs / outputs\n",
      "    a. The whole genome sequence information from our experimental group we want to make an alternate reference for. This can either be FASTQ, compatible VCF or BCF files, etc.\n",
      "    b. The 'common' reference genome we are going to convolute. \n",
      "    c. (optional) SNP priors. \n",
      "    d. (optional) Structural variation priors (indels, splices, de novo contigs) \n",
      "    _. (future) Epigenetic variation priors (is there common epigenetics? probably)\n",
      "    da. (optional) Consensus priors\n",
      "    e. Output filename for new reference. \n",
      "    f. Output filename for liftover file\n",
      "    g. (optional) Output filename for log (defaults to stderr)\n",
      "\n",
      "\n",
      "1. Pre-processing: Align the whole genomes\n",
      "    -if priors, generate new consensus\n",
      "    Stampy!\n",
      "        - get list of unmapped reads\n",
      "    - prestatistics calculation\n",
      "\n",
      "2. Variant calling on whole genomes\n",
      "    GATK\n",
      "\n",
      "3. SV calling / CNV on whole genomes\n",
      "    GATK\n",
      "    MapSplice\n",
      "    Contra CNV\n",
      "\n",
      "4. De novo alignment of unmapped reads\n",
      "    cortex-var\n",
      "\n",
      "5. Attempt to place new contigs on genome\n",
      "    ? Does this exist ?\n",
      "    Structure / MALDsoft\n",
      "    - output: variation file\n",
      "\n",
      "6. Make alternate sequence\n",
      "    GATK\n",
      "\n",
      "7. Realignment to alternate sequence\n",
      "    Stampy\n",
      "\n",
      "8. validation step here?\n",
      "\n",
      "9. Results\n",
      "\n",
      "Subroutines:\n",
      "    Make alternate reference from priors\n",
      "        Takes: \n",
      "            SNP priors\n",
      "            SV priors\n",
      "            Epi priors\n",
      "            human consensus / artificial consensus that the priors are coded in\n",
      "        Outputs\n",
      "            new consensus sequence\n",
      "            liftover file describing changes from one to the next\n",
      "    Liftover\n",
      "        Takes\n",
      "            change file\n",
      "            old reference\n",
      "            (optional) new reference\n",
      "        Produces\n",
      "            chain liftover file\n",
      "    Validation step"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**Things to think about**\n",
      "\n",
      "1. Do we preserve the error / entropy of the model when we convolute the variants with our common variance?\n",
      "2. De-novo realignment in regions where there is little consensus to find large SVs\n",
      "    * Has GATK included this in their new tool?\n",
      "3. Incorporating Bayesian statistics in calculating what a 'common' variant is / isn't. It should only make it common if there is enough data to support it. \n",
      "4. *Important* We need to make sure there is a SNP translation file made.\n",
      "5. MapReduce from the beginning? YES\n",
      "6. **Validation:** take new reference from 25% of samples, test on 75% of samples"
     ]
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "Python stuff:\n",
      "    Bcbio-nextgen\n",
      "    CLEVER (CLique Enumerating Variant findER)\n",
      "    Contra CNV\n",
      "    CRISP multisample variant caller for pooled data\n",
      "    DeepTools visualization\n",
      "    MapSplice for detecting alternate splicing\n",
      "    NGSUtils\n",
      "    Pybedtools\n",
      "    Stampy hybrid mapper***\n",
      "    \n",
      "\n",
      "Other:\n",
      "    Cortex-var"
     ]
    }
   ],
   "metadata": {}
  }
 ]
}