{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Python-scarecrow test code (working copy)"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Imports / style (run this first always)\n",
      "\n",
      "%matplotlib inline\n",
      "from collections import defaultdict\n",
      "import json\n",
      "import sys\n",
      "import time\n",
      "\n",
      "import numpy as np\n",
      "import scipy as sp\n",
      "import matplotlib.pyplot as plt\n",
      "import pandas as pd\n",
      "\n",
      "from matplotlib import rcParams\n",
      "import matplotlib.cm as cm\n",
      "import matplotlib as mpl\n",
      "\n",
      "class AwesomeError(Exception):\n",
      "     def __init__(self, value):\n",
      "         self.value = value\n",
      "         pass\n",
      "     def __str__(self):\n",
      "         return repr(self.value)\n",
      "         pass\n",
      "\n",
      "#colorbrewer2 Dark2 qualitative color table\n",
      "dark2_colors = [(0.10588235294117647, 0.6196078431372549, 0.4666666666666667),\n",
      "                (0.8509803921568627, 0.37254901960784315, 0.00784313725490196),\n",
      "                (0.4588235294117647, 0.4392156862745098, 0.7019607843137254),\n",
      "                (0.9058823529411765, 0.1607843137254902, 0.5411764705882353),\n",
      "                (0.4, 0.6509803921568628, 0.11764705882352941),\n",
      "                (0.9019607843137255, 0.6705882352941176, 0.00784313725490196),\n",
      "                (0.6509803921568628, 0.4627450980392157, 0.11372549019607843)]\n",
      "\n",
      "rcParams['figure.figsize'] = (10, 6)\n",
      "rcParams['figure.dpi'] = 150\n",
      "rcParams['axes.color_cycle'] = dark2_colors\n",
      "rcParams['lines.linewidth'] = 2\n",
      "rcParams['axes.facecolor'] = 'white'\n",
      "rcParams['font.size'] = 14\n",
      "rcParams['patch.edgecolor'] = 'white'\n",
      "rcParams['patch.facecolor'] = dark2_colors[0]\n",
      "rcParams['font.family'] = 'StixGeneral'\n",
      "\n",
      "\n",
      "def remove_border(axes=None, top=False, right=False, left=True, bottom=True):\n",
      "    \"\"\"\n",
      "    Minimize chartjunk by stripping out unnecesasry plot borders and axis ticks\n",
      "    \n",
      "    The top/right/left/bottom keywords toggle whether the corresponding plot border is drawn\n",
      "    \"\"\"\n",
      "    ax = axes or plt.gca()\n",
      "    ax.spines['top'].set_visible(top)\n",
      "    ax.spines['right'].set_visible(right)\n",
      "    ax.spines['left'].set_visible(left)\n",
      "    ax.spines['bottom'].set_visible(bottom)\n",
      "    \n",
      "    #turn off all ticks\n",
      "    ax.yaxis.set_ticks_position('none')\n",
      "    ax.xaxis.set_ticks_position('none')\n",
      "    \n",
      "    #now re-enable visibles\n",
      "    if top:\n",
      "        ax.xaxis.tick_top()\n",
      "    if bottom:\n",
      "        ax.xaxis.tick_bottom()\n",
      "    if left:\n",
      "        ax.yaxis.tick_left()\n",
      "    if right:\n",
      "        ax.yaxis.tick_right()\n",
      "        \n",
      "pd.set_option('display.width', 500)\n",
      "pd.set_option('display.max_columns', 100)\n",
      "\n",
      "import Bio as bp\n",
      "from Bio.Sequencing.Applications import BwaAlignCommandline as bwa_aln\n",
      "from Bio.Sequencing.Applications import BwaSamseCommandline as bwa_samse\n",
      "from Bio.Sequencing.Applications import BwaSampeCommandline as bwa_sampe\n",
      "from Bio.Sequencing.Applications import BwaIndexCommandline as bwa_index\n",
      "from Bio.Sequencing.Applications import BwaBwaswCommandline as bwa_bwasw\n",
      "import HTSeq as ht\n",
      "import subprocess"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#colorbrewer2 Dark2 qualitative color table\n",
      "dark2_colors = [(0.10588235294117647, 0.6196078431372549, 0.4666666666666667),\n",
      "                (0.8509803921568627, 0.37254901960784315, 0.00784313725490196),\n",
      "                (0.4588235294117647, 0.4392156862745098, 0.7019607843137254),\n",
      "                (0.9058823529411765, 0.1607843137254902, 0.5411764705882353),\n",
      "                (0.4, 0.6509803921568628, 0.11764705882352941),\n",
      "                (0.9019607843137255, 0.6705882352941176, 0.00784313725490196),\n",
      "                (0.6509803921568628, 0.4627450980392157, 0.11372549019607843)]\n",
      "\n",
      "rcParams['figure.figsize'] = (10, 6)\n",
      "rcParams['figure.dpi'] = 150\n",
      "rcParams['axes.color_cycle'] = dark2_colors\n",
      "rcParams['lines.linewidth'] = 2\n",
      "rcParams['axes.facecolor'] = 'white'\n",
      "rcParams['font.size'] = 14\n",
      "rcParams['patch.edgecolor'] = 'white'\n",
      "rcParams['patch.facecolor'] = dark2_colors[0]\n",
      "rcParams['font.family'] = 'StixGeneral'\n",
      "\n",
      "\n",
      "def remove_border(axes=None, top=False, right=False, left=True, bottom=True):\n",
      "    \"\"\"\n",
      "    Minimize chartjunk by stripping out unnecesasry plot borders and axis ticks\n",
      "    \n",
      "    The top/right/left/bottom keywords toggle whether the corresponding plot border is drawn\n",
      "    \"\"\"\n",
      "    ax = axes or plt.gca()\n",
      "    ax.spines['top'].set_visible(top)\n",
      "    ax.spines['right'].set_visible(right)\n",
      "    ax.spines['left'].set_visible(left)\n",
      "    ax.spines['bottom'].set_visible(bottom)\n",
      "    \n",
      "    #turn off all ticks\n",
      "    ax.yaxis.set_ticks_position('none')\n",
      "    ax.xaxis.set_ticks_position('none')\n",
      "    \n",
      "    #now re-enable visibles\n",
      "    if top:\n",
      "        ax.xaxis.tick_top()\n",
      "    if bottom:\n",
      "        ax.xaxis.tick_bottom()\n",
      "    if left:\n",
      "        ax.yaxis.tick_left()\n",
      "    if right:\n",
      "        ax.yaxis.tick_right()\n",
      "        \n",
      "pd.set_option('display.width', 500)\n",
      "pd.set_option('display.max_columns', 100)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from multiprocessing import Process, Manager\n",
      "\n",
      "def f(d):\n",
      "    d[1] += '1'\n",
      "    d['2'] += 2\n",
      "\n",
      "if __name__ == '__main__':\n",
      "    manager = Manager()\n",
      "\n",
      "    d = manager.dict()\n",
      "    d[1] = '1'\n",
      "    d['2'] = 2\n",
      "\n",
      "    p1 = Process(target=f, args=(d,))\n",
      "    p2 = Process(target=f, args=(d,))\n",
      "    p1.start()\n",
      "    p2.start()\n",
      "    p1.join()\n",
      "    p2.join()\n",
      "\n",
      "    print d"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "{1: '111', '2': 6}\n"
       ]
      }
     ],
     "prompt_number": 13
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "set_cells = set()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 23
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "set_cells = set([783, 1086, 1425, 1416, 1323, 1416, 351, 1108, 995, 1425, 155, 390, 1086, 783, 351, 668, 1105, 472, 668, 1105, 1416, 351, 1060, 1108, 155, 1507, 1323, 1449, 1507, 995, 1425, 783])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 25
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(set_cells)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 26,
       "text": [
        "16"
       ]
      }
     ],
     "prompt_number": 26
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "'''\n",
      "scarecrow\n",
      "\n",
      "Usage:\n",
      "    Make alternate reference sequences\n",
      "\n",
      "Inputs:\n",
      "    seq\n",
      "        A filename to the fastq sequence that the alternate genome should be made from\n",
      "    ref\n",
      "        the consensus sequence that should be used\n",
      "    prior_snp\n",
      "        (optional) SNP priors\n",
      "    prior_sv\n",
      "        (optional) structural variation priors\n",
      "    prior_con\n",
      "        (optional) prior consensus sequence\n",
      "\n",
      "Returns:\n",
      "    new_ref\n",
      "        a tuple containing the .fa alternate consensus file and its index\n",
      "    applied_variants\n",
      "        a DataFrame containing the applied changes between the current reference and the new one\n",
      "    liftover\n",
      "        a chain file to convert between the two references\n",
      "    error_log\n",
      "        text containing errors and debugging information\n",
      "\n",
      "'''\n",
      "\n",
      "import Bio as bp\n",
      "from Bio.Sequencing.Applications import BwaAlignCommandline as bwa_aln\n",
      "from Bio.Sequencing.Applications import BwaSamseCommandline as bwa_samse\n",
      "from Bio.Sequencing.Applications import BwaSampeCommandline as bwa_sampe\n",
      "from Bio.Sequencing.Applications import BwaIndexCommandline as bwa_index\n",
      "from Bio.Sequencing.Applications import BwaBwaswCommandline as bwa_bwasw\n",
      "import HTSeq\n",
      "\n",
      "seq1 = '/media/ryan/RyanNeffBioinformatics/phenom/Data/raw_seq/SRR350098_1.filt.fastq'\n",
      "seq2 = '/media/ryan/RyanNeffBioinformatics/phenom/Data/raw_seq/SRR350098_2.filt.fastq'\n",
      "ref = '/media/ryan/RyanNeffBioinformatics/phenom/Data/references/hg19.fa'\n",
      "\n",
      "def scarecrow(seq1='', seq2='', ref='',prior_snp='', prior_sv='', prior_con=''):\n",
      "    # step 1 - align the sequence data to the reference provided\n",
      "    sai_file1='bwa_aln_1.sai'\n",
      "    sai_file2='bwa_aln_2.sai'\n",
      "    bwa_cmd = bwa_aln(reference=ref, read_file=seq1, R=100, q=10, t=8)\n",
      "    bwa_cmd(sai_file1)\n",
      "    bwa_cmd = bwa_aln(reference=ref, read_file=seq2, R=100, q=10, t=8)\n",
      "    bwa_cmd(sai_file2)\n",
      "    bwa_cmd = bwa_sampe(sai_file1=sai_file2, sai_file2=sai_file2,\n",
      "                        read_file1=seq1, read_file2=seq2)\n",
      "    \n",
      "    # step 2 - variant calling\n",
      "    # step 3 - SV calling\n",
      "    # step 4 - CNV calling\n",
      "    # step 5 - de novo alignment of unmapped reads\n",
      "    # step 6 - attempt to place new unmapped contigs generated in step 5 onto the genome\n",
      "    # step 7 - filter ALL changes to genome somehow (50% for now)\n",
      "    # step 8 - make alternate reference\n",
      "    # step 9 - (?) Realignment to alternate sequence (cross-validation?)\n",
      "    # step 10 - format for return\n",
      "    return new_ref, applied_variants, liftover, error_log"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "bwa_cmd = bwa_aln(reference=ref, read_file=seq1, R=100)\n",
      "print(bwa_cmd)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "bwa aln /media/ryan/RyanNeffBioinformatics/phenom/Data/references/hg19.fa /media/ryan/RyanNeffBioinformatics/phenom/Data/raw_seq/SRR350098_1.filt.fastq -R 100\n"
       ]
      }
     ],
     "prompt_number": 6
    }
   ],
   "metadata": {}
  }
 ]
}